onnx/compose.py

```diff
@@ -0,0 +1,313 @@
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from typing import List, Tuple, Text, Optional, MutableMapping
+from onnx import ModelProto, GraphProto, helper, checker
+from onnx import TensorProto as tp
+
+
+def merge_graphs(
+        g1,  # type: GraphProto
+        g2,  # type: GraphProto
+        io_map,  # type: List[Tuple[Text, Text]]
+        name=None,  # type: Optional[Text]
+        doc_string=None,  # type: Optional[Text]
+):  # type: (...) -> GraphProto
+    """Combines two ONNX graphs into a single one.
+
+    The combined graph is defined by connecting the specified set of outputs/inputs. Those inputs/outputs
+    not specified in the io_map argument will remain as inputs/outputs of the combined graph.
+
+    Arguments:
+        g1 (GraphProto): First graph
+        g2 (GraphProto): Second graph
+        io_map (list of pairs of string): The pairs of names [(out0, in0), (out1, in1), ...]
+                                          representing outputs of the first graph and inputs of the second
+                                          to be connected
+        name (string): Optional name for the combined graph
+                       By default, the name is g1.name and g2.name concatenated with an undescore delimiter
+        doc_string (string): Optional docstring for the combined graph
+                             If not provided, a default docstring with the concatenation of g1 and g2 docstrings is used
+    """
+    if type(g1) is not GraphProto:
+        raise ValueError("g1 argument is not an ONNX graph")
+    if type(g2) is not GraphProto:
+        raise ValueError("g2 argument is not an ONNX graph")
+
+    io_map_g1_outs = [io[0] for io in io_map]
+    io_map_g2_ins = [io[1] for io in io_map]
+    g1_outs = [o.name for o in g1.output]
+    g2_ins = [i.name for i in g2.input]
+
+    for g1_out_name, g2_in_name in io_map:
+        if g1_out_name not in g1_outs:
+            raise ValueError(f"Output {g1_out_name} not present in g1")
+        if g2_in_name not in g2_ins:
+            raise ValueError(f"Input {g2_in_name} not present in g2")
+
+    g = GraphProto()
+
+    g.node.extend(g1.node)
+    g.node.extend(g2.node)
+
+    # Connecting outputs of the first graph with the inputs of the second
+    for g1_out_name, g2_in_name in io_map:
+        for node in g.node:
+            for index, name in enumerate(node.input):
+                if name == g2_in_name:
+                    node.input[index] = g1_out_name
+
+    g.input.extend(g1.input)
+    g.input.extend(
+        [item for item in g2.input if item.name not in io_map_g2_ins])
+
+    g.output.extend(
+        [item for item in g1.output if item.name not in io_map_g1_outs])
```

Comment:
This looks like a reasonable heuristic. However, I wonder if it would be useful to allow users to override this by explicitly indicating which outputs should be included as outputs of the merged-graph?

Reply:
Done. Added optional `inputs` and `outputs` argument, with the help extract_model utility